### PR TITLE
Irreps improvements

### PR DESCRIPTION
**Pull request by bdice**
## Description
- Minor improvements to Irrep/Irreps docs and implementation.
- Add test for empty Irreps.

This started as an attempt to change `Irrep` / `Irreps` so that they weren't subclassed from `tuple`, but that failed because of issues with TorchScript. These classes are too complex to be written as TorchScript-compatible classes, so instead I pivoted to making some minor improvements to the class's docs, implementation, and testing.

## Motivation and Context
Additional context in PR comments.

## How Has This Been Tested?
Existing tests pass, new tests added.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds or improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation improvement (updates to user guides, docstrings, or developer docs)

## Checklist:
- [x] I have read the [**CONTRIBUTING**](https://github.com/e3nn/e3nn/blob/main/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [x] I have updated the documentation (if relevant).
- [x] I have added tests that cover my changes (if relevant).
- [x] All new and existing tests passed.
- [x] I have updated the [Changelog](https://github.com/e3nn/e3nn/blob/main/ChangeLog.md).

### CODE REVIEW

File: e3nn/o3/_irreps.py [hunk 1]
```diff
@@ -14,17 +15,14 @@ class Irrep(tuple):
...
+    Parameters
     ----------
     l : int
-        non negative integer, the degree of the representation, :math:`l = 0, 1, \dots`
+        non-negative integer, the degree of the representation, :math:`l = 0, 1, \dots`
 
     p : {1, -1}
         the parity of the representation
 
-    dim : int
```
Comment (bdice) [inline e3nn/o3/_irreps.py:25]: This was missing a "Parameters" section, so I renamed this. I added a docstring to the properties, instead.

File: e3nn/o3/_irreps.py [hunk 2]
```diff
@@ -52,29 +50,43 @@ class Irrep(tuple):
...
+
+    @property
+    def p(self) -> int:
+        r"""The parity of the representation, :math:`p = \pm 1`."""
+        return self[1]
+
+    def __iter__(self):
+        yield self.l
+        yield self.p
+
```
Comment (bdice) [inline e3nn/o3/_irreps.py:89]: Remove extraneous change.
```suggestion
```

File: e3nn/o3/_irreps.py [hunk 3]
```diff
@@ -353,13 +359,13 @@ def __new__(cls, irreps=None):
...
-                    warnings.warn("prefer using [(mul, (l, p))] to distinguish multiplicity from irrep", DeprecationWarning, stacklevel=2)
+                    warnings.warn("Use the argument format [(mul, (l, p)), ...] to distinguish multiplicity from irrep", DeprecationWarning, stacklevel=2)
                 else:
                     mul = None
                     ir = None
 
-                assert isinstance(mul, int) and mul >= 0
-                assert ir is not None
+                if not (isinstance(mul, int) and mul >= 0 and ir is not None):
+                    raise ValueError(f"Unable to interpret \"{mul_ir}\" as an irrep.")
```
Comment (bdice) [inline e3nn/o3/_irreps.py:368]: Raising a `ValueError` here is probably more Pythonic than an assertion. Assertions can be disabled in "optimized" Python and aren't as informative as an exception like `ValueError`, `TypeError`, etc. I did this in a couple other places, just wanted to explain why. 👍

File: e3nn/o3/_irreps.py [hunk 4]
```diff
@@ -423,7 +429,8 @@ def randn(self, *size, normalization='component', requires_grad=False, dtype=Non
         >>> Irreps("5x0e + 10x1o").randn(5, -1, 5, normalization='norm').shape
         torch.Size([5, 35, 5])
 
-        >>> Irreps("2o").randn(2, -1, 3, normalization='norm').norm(dim=1).sub(1).abs().max().item() < 1e-5
+        >>> random_tensor = Irreps("2o").randn(2, -1, 3, normalization='norm')
+        >>> random_tensor.norm(dim=1).sub(1).abs().max().item() < 1e-5
```
Comment (bdice) [inline e3nn/o3/_irreps.py:433]: This rendered a little funny in the docs because the line was too long, so I split it.

File: e3nn/o3/_irreps.py [hunk 5]
```diff
@@ -607,9 +613,6 @@ def D_from_matrix(self, R):
         R : `torch.Tensor`
             tensor of shape :math:`(..., 3, 3)`
 
-        k : `torch.Tensor`, optional
```
Comment (bdice) [inline e3nn/o3/_irreps.py:610]: This isn't a parameter of the function so I removed it.

**Review by bdice — COMMENTED** [review]
Comments for context.

File: e3nn/o3/_irreps.py [hunk 6]
```diff
@@ -52,29 +50,39 @@ class Irrep(tuple):
...
+            elif isinstance(l, tuple):
+                l, p = l
 
         assert isinstance(l, int) and l >= 0, l
         assert p in [-1, 1], p
         return super().__new__(cls, (l, p))
 
+    @property
+    def l(self) -> int:
+        r"""The dimension of the representation, :math:`l = 0, 1, \dots`."""
```
Comment (mariogeiger) [inline e3nn/o3/_irreps.py:78]: Typo here
Dimension-> degree

**Review by mariogeiger — COMMENTED** [review]
Thanks!
